tsdb/docs/refs.md

```diff
@@ -0,0 +1,82 @@
+Once mmapped, chunks have HeadSeriesRefs inside them, allowing you to recreate the index from reading chunks
+(along with WAL which has the labels. it also has all those points, but by using chunks we can save cpu/time and not replay all of WAL on startup)
+
+2) Due to an unrelated hack (TODO clarify), you may not use values >= uint32 inside the uint64's. so we presume prometheus gets restarted before you reach 2^32 seriesRefs in head, but don't check this.
+the last series ref is always replayed from the WAL and is continued from there.
+3) During querying, HeadSeriesRef are limited to 2^40 (see HeadChunkRef)
+
+## BlockSeriesRef
+
+Persistent blocks are all independent entities and the format/structure is completely different from head block.
+
+In blocks, series are lexicographically ordered by labels and the byte offset in the index file (divided by 16 because they're all aligned on 16 bytes) becomes the BlockSeriesRef.
+They are not sequential because index entries may be multiples of 16 bytes. And they don't start from 0 because the byte offset is absolute and includes the magic number, symbols table, etc.
+BlockSeriesRef. are only 32 bits, because 64 bits would slow down the postings lists disk access. (note: this limits the index size to 2^32 * 16 = 64 GB)
+
+See also:
+* https://ganeshvernekar.com/blog/prometheus-tsdb-persistent-block-and-its-index/#3-index
+* https://ganeshvernekar.com/blog/prometheus-tsdb-persistent-block-and-its-index/#c-series
+
+## ChunkRef
+
+Note: we cover the implementations as used in Prometheus.  Other projects may use different implementations.
+
+### HeadChunkRef
+
+A HeadChunkRef is an 8 byte integer, that is only used during querying.  It packs together:
+
+* 5 Bytes for HeadSeriesRef.
+* 3 Bytes for ChunkID (uint64). This is simply an index into a slice of mmappedChunks for a given series
+
+There are two implications here:
+
+* While HeadSeriesRefs can during ingestion go higher, during querying they are limited to 2^40.  Querying too high numbers will lead to query failures (but not impact ingestion).
+* ChunkID keeps growing as we enter new chunks. If prometheus runs too long, we might hit 2^24. If id=len(mmappedchunks) then it's the head chunk. note.
+
+### BlockChunkRef
+
+A BlockChunkRef is an 8 byte integer
+TODO is it only used during querying like HeadChunkRef?
+
+It packs together:
+
+* 4 Bytes for chunk file index in the block. This number just increments. Filenames [start at 1](https://ganeshvernekar.com/blog/prometheus-tsdb-persistent-block-and-its-index/#contents-of-a-block) but the references probably start at 0 (TODO confirm)
+* 4 Bytes offset within the file
+
+### Why does HeadChunkRef contain a series reference and BlockChunkRef does not?
+
+In head the index entry has both series and data hence a chunkref includes a seriesref
+In a block, series and chunk data are separate files.  A typical query gets resolved like so:
```

Comment:
I don't quite get this.

Reply:
I feel both explanations can be not very clear. I have a suggestion here

```
The aim of both the ChunkRef is to get to the chunk data as efficiently as possible; it gives us just enough information to get to the chunk.
In case of the Head block, the chunks are in the series struct. So we need to reach the series before we can access the chunk from it. Hence we need to pack the `HeadSeriesRef` to get to the series.
In case of persistent block, the chunk files are separated from the index and static. Hence you only need the co-ordinates within the `chunks` directory to get to the chunk. Hence no need of `BlockSeriesRef`.
```

Reply:
that makes most sense. will fix